File: perfect-numbers/src/perfect_numbers.erl
-module(perfect_numbers).

-export([classify/1]).

test_class(Number,Number) ->
    perfect;
test_class(Number,AliquotSum) when Number < AliquotSum ->
    abundant;
test_class(Number,AliquotSum) when Number > AliquotSum ->
    deficient.


classify(Number) when Number > 2 ->
    Seq = lists:seq(1, ( Number div 2 ) + 1),
    DivisorSeq = [X || X <- Seq, Number rem X == 0],
    AliquotSum = lists:sum(DivisorSeq),
    test_class(Number,AliquotSum);
classify(2) ->
    deficient;
classify(1) ->
    deficient.

File: perfect-numbers/lib/perfect_numbers.ex
defmodule PerfectNumbers do
  @doc """
  Determine the aliquot sum of the given `number`, by summing all the factors
  of `number`, aside from `number` itself.

  Based on this sum, classify the number as:

  :perfect if the aliquot sum is equal to `number`
  :abundant if the aliquot sum is greater than `number`
  :deficient if the aliquot sum is less than `number`
  """
  @spec classify(number :: integer) :: {:ok, atom} | {:error, String.t()}
  def classify(number) when number < 1 do
    {:error, "Classification is only possible for natural numbers."}
  end
  def classify(1) do
    {:ok, :deficient}
  end
  def classify(number) when is_integer(number) do
    aliquot_sum = 1..(div(number, 2)) |> Stream.filter(&(rem(number, &1) == 0)) |> Enum.sum()

    case aliquot_sum do
      ^number -> {:ok, :perfect}
      aliquot_sum when aliquot_sum > number -> {:ok, :abundant}
      aliquot_sum when aliquot_sum < number -> {:ok, :deficient}
    end
  end
end
